### PR TITLE
TAJO-680: Improve the IN operator to support sub queries.

### PR DESCRIPTION
**Pull request by jihoonson**
This patch is based on #593, so please review it first.

This work includes:
* Improving the IN expression to allow uncorrelated subquery
 * Throwing an exception in the case of correlated subquery
* Improving the join ordering to allow semi and anti joins
* Set of query result tests

This work does not include the query plan tests because it makes the patch size huge.
I'll add them at https://issues.apache.org/jira/browse/TAJO-1671.

[Review on Reviewable](https://reviewable.io/reviews/apache/tajo/620)

### CODE REVIEW

**Comment by hyunsik** [comment]
Hi,

I'm going to review this work. Could you rebase it against the latest master branch?

**Comment by jihoonson** [comment]
Hi @hyunsik. I've rebased.

File: tajo-core/src/test/java/org/apache/tajo/engine/query/TestInSubquery.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  @SimpleTest()
+  public final void testSameKeyNameOfOuterAndInnerQueries() throws Exception {
+    runSimpleTests();
+  }
+
+  @Test
+  @Option(parameterized = true, sort = true)
+  @SimpleTest()
+  public final void testWithAsteriskAndJoin() throws Exception {
+    // select * from lineitem, orders where l_orderkey = o_orderkey and l_partkey in (select l_partkey from lineitem where l_linenumber in (1, 3, 5, 7, 9))
```
Comment (hyunsik) [inline tajo-core/src/test/java/org/apache/tajo/engine/query/TestInSubquery.java]: It exceeds the column width of coding standards.

File: tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/JoinOrderingUtil.java [hunk 2]
```diff
@@ -139,6 +141,11 @@ public static boolean isAssociativeJoin(JoinGraphContext context, JoinEdge leftE
    * (A full B) full C    | A full (B full C)     | Equivalent
    * ==============================================================
    *
+<<<<<<< HEAD
```
Comment (hyunsik) [inline tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/JoinOrderingUtil.java]: You may miss this line which seems to be inserted by git merge phase.

File: tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/JoinOrderingUtil.java [hunk 3]
```diff
@@ -139,6 +141,11 @@ public static boolean isAssociativeJoin(JoinGraphContext context, JoinEdge leftE
...
+   * Cross, Semi and Anti joins follow the rule of the Inner join.
+   *
+=======
+>>>>>>> c50a5dadff90fa90709abbce59856e834baa4867
```
Comment (hyunsik) [inline tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/JoinOrderingUtil.java]: You may miss this line which seems to be inserted by git merge phase.

File: tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/InSubqueryRewriteRule.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.tajo.plan.rewrite.rules;
+
+import com.google.common.base.Preconditions;
+import org.apache.tajo.OverridableConf;
```
Comment (hyunsik) [inline tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/InSubqueryRewriteRule.java]: unused import

**Comment by hyunsik** [comment]
+1
This patch looks great to me, and it includes enough unit tests to verify its feature. I leaved very trivial comments. You can commit it after they resolve them.

**Comment by jihoonson** [comment]
Thank you for the review! I've committed after changing the patch according to your comment.

**Pull request closed without merging**